### PR TITLE
LSM6DSO accel/gyro sensor: Add DTS property for range/odr/power-mode

### LINKED ISSUE #41117: Incorrect setting of gyro sensitivity in LSM6DSO driver
**Describe the bug**

The LSM6DSO driver does not set the gyro sensitivity (full-scale setting) correctly. An example walk-through of the issue when trying to set ±2000°/s full-scale:

Zephyr driver gyro configuration (hard-coded magic numbers that don't align with stmemsc API):

https://github.com/zephyrproject-rtos/zephyr/blob/13bf2e5e48737f5971bdfa2ee683453e379403ea/drivers/sensor/lsm6dso/lsm6dso.h#L67-L86

`LSM6DSO_DEFAULT_GYRO_FULLSCALE` used:

https://github.com/zephyrproject-rtos/zephyr/blob/9407add27c16f5a10eb1b508f8cb6a482bfc62ed/drivers/sensor/lsm6dso/lsm6dso.c#L782

Magic numbers passed through directly into stmemsc API:

https://github.com/zephyrproject-rtos/zephyr/blob/9407add27c16f5a10eb1b508f8cb6a482bfc62ed/drivers/sensor/lsm6dso/lsm6dso.c#L136-L146

Expected values for `lsm6dso_gy_full_scale_set` function call:

https://github.com/zephyrproject-rtos/hal_st/blob/575de9d461aa6f430cf62c58a053675377e700f3/sensor/stmemsc/lsm6dso_STdC/driver/lsm6dso_reg.h#L2775-L2784

The stmemsc API is such that the value of the `lsm6dso_fs_g_t` enum is directly used to set the `FS[1:0]_G` and `FS_125` fields in the `CTRL2_G` register. The values that the Zephyr driver are passing it appear to bear no relation to the `lsm6dso_fs_g_t` values and the end result is that, when requesting ±2000°/s full scale the gyro is actually configured for ±125°/s full scale:

 1. `LSM6DSO_DEFAULT_GYRO_FULLSCALE` == 3
 2. `lsm6dso_gy_full_scale_set` sets the `fs_g` member of the `lsm6dso_ctrl2_g_t` struct, which effectively results in a left shift by 1, setting the lower nibble to 6 (1010 in binary). Since the `FS_125` bit is set, the value of the `FS[1:0]_G` field is ignored by the IMU.

### PR DESCRIPTION
**Pull request by avisconti**
LSM6DSO sensor (both accel and gyro)                                                                              
                                        
Move odr/range properties from Kconfig to DTS. Moreover add power-mode properties as well.                                                                                
Fixes #41117

### CODE REVIEW

File: dts/bindings/sensor/st,lsm6dso-common.yaml [hunk 1]
```diff
@@ -18,9 +19,98 @@ properties:
...
+        - 0 # High Performance mode (default)
+        - 1 # Low/Normal Power mode
+
+    gyro-range:
+      type: int
+      required: false
+      default: 0
+      description: |
+        Range in dps. Default is power-up configuration.
+      enum:
```
Comment (richardbarlow) [inline dts/bindings/sensor/st,lsm6dso-common.yaml:91]: All of the full-scale ranges should be degrees per seconds and all of the resolutions should be in millidegrees per second.

Comment (avisconti) [reply dts/bindings/sensor/st,lsm6dso-common.yaml:91]: > All of the full-scale ranges should be degrees per seconds and all of the resolutions should be in millidegrees per second. (In the comments)

OK, thanks for catching it!!

File: drivers/sensor/lsm6dso/lsm6dso.c [hunk 2]
```diff
@@ -766,28 +744,35 @@ static int lsm6dso_init_chip(const struct device *dev)
 
 	k_busy_wait(100);
 
-	if (lsm6dso_accel_set_fs_raw(dev,
-				     LSM6DSO_DEFAULT_ACCEL_FULLSCALE) < 0) {
-		LOG_DBG("failed to set accelerometer full-scale");
+	fs = cfg->accel_range;
+	LOG_DBG("accel range is %d", fs);
+	if (lsm6dso_accel_set_fs_raw(dev, fs) < 0) {
+		LOG_DBG("failed to set accelerometer full-scale %d", fs);
```
Comment (MaureenHelm) [inline drivers/sensor/lsm6dso/lsm6dso.c:765]: This should be `LOG_ERR`. It's also a little confusing that you call it "accelerometer full-scale" here, but "accel range" two lines above.

File: drivers/sensor/lsm6dso/lsm6dso.c [hunk 3]
```diff
@@ -766,28 +744,35 @@ static int lsm6dso_init_chip(const struct device *dev)
...
 		return -EIO;
 	}
-	lsm6dso->acc_gain = LSM6DSO_DEFAULT_ACCEL_SENSITIVITY;
+	lsm6dso->acc_gain = (lsm6dso_accel_fs_sens[fs] * GAIN_UNIT_XL);
```
Comment (MaureenHelm) [inline drivers/sensor/lsm6dso/lsm6dso.c:768]: Is there a reason you wrapped the expression with parentheses?

File: drivers/sensor/lsm6dso/lsm6dso.c [hunk 4]
```diff
@@ -766,28 +744,35 @@ static int lsm6dso_init_chip(const struct device *dev)
...
 
-	lsm6dso->accel_freq = lsm6dso_odr_to_freq_val(CONFIG_LSM6DSO_ACCEL_ODR);
-	if (lsm6dso_accel_set_odr_raw(dev, CONFIG_LSM6DSO_ACCEL_ODR) < 0) {
-		LOG_DBG("failed to set accelerometer sampling rate");
+	odr = cfg->accel_odr;
+	LOG_DBG("accel odr is %d", odr);
+	lsm6dso->accel_freq = lsm6dso_odr_to_freq_val(odr);
+	if (lsm6dso_accel_set_odr_raw(dev, odr) < 0) {
+		LOG_DBG("failed to set accelerometer sampling rate %d", odr);
```
Comment (MaureenHelm) [inline drivers/sensor/lsm6dso/lsm6dso.c:774]: This should be `LOG_ERR`. It's also a little confusing that you call it "accelerometer sampling rate" here, but "accel odr" two lines above.

File: drivers/sensor/lsm6dso/lsm6dso.c [hunk 5]
```diff
@@ -766,28 +744,35 @@ static int lsm6dso_init_chip(const struct device *dev)
...
 		return -EIO;
 	}
 
-	if (lsm6dso_gyro_set_fs_raw(dev, LSM6DSO_DEFAULT_GYRO_FULLSCALE) < 0) {
-		LOG_DBG("failed to set gyroscope full-scale");
+	fs = cfg->gyro_range;
+	LOG_DBG("gyro range is %d", fs);
+	if (lsm6dso_gyro_set_fs_raw(dev, fs) < 0) {
+		LOG_DBG("failed to set gyroscope full-scale %d", fs);
```
Comment (MaureenHelm) [inline drivers/sensor/lsm6dso/lsm6dso.c:793]: ditto

File: drivers/sensor/lsm6dso/lsm6dso.c [hunk 6]
```diff
@@ -766,28 +744,35 @@ static int lsm6dso_init_chip(const struct device *dev)
...
+	lsm6dso->gyro_gain = (lsm6dso_gyro_fs_sens[fs] * GAIN_UNIT_G);
 
-	lsm6dso->gyro_freq = lsm6dso_odr_to_freq_val(CONFIG_LSM6DSO_GYRO_ODR);
-	if (lsm6dso_gyro_set_odr_raw(dev, CONFIG_LSM6DSO_GYRO_ODR) < 0) {
-		LOG_DBG("failed to set gyroscope sampling rate");
+	odr = cfg->gyro_odr;
+	LOG_DBG("gyro odr is %d", odr);
+	lsm6dso->gyro_freq = lsm6dso_odr_to_freq_val(odr);
+	if (lsm6dso_gyro_set_odr_raw(dev, odr) < 0) {
+		LOG_DBG("failed to set gyroscope sampling rate %d", odr);
```
Comment (MaureenHelm) [inline drivers/sensor/lsm6dso/lsm6dso.c:802]: ditto

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/sensor/lsm6dso/lsm6dso.c (hunk 2 above)
Comment (avisconti) [reply drivers/sensor/lsm6dso/lsm6dso.c:765]: ACK to this one and the rest of the observations here below as well. Thanks!

**Comment by avisconti** [comment]
@mbolivar-nordic can you please verify whether the dts binding policy is satisfied here? It should be so, but just in case...

**Review by gmarull — APPROVED** [review]

**Review by MaureenHelm — APPROVED** [review]

**Comment by richardbarlow** [comment]
I've just tested this on our hardware and it functions as expected. We are now able to configure the gyro in the ±2000dps range. Thank you @avisconti for the very quick fix for #41117.

**Pull request merged**